### PR TITLE
Misc optimizations

### PR DESCRIPTION
**Pull request by alcuadrado**
Sixth PR of the compilation times optimization effort.

This PR introduces multiple smaller optimizations:

* I created an fs-util module to replace slower dependencies with it (e.g. `glob`)
* Stopped using `loadash` whenever possible, and only load its necessary modules when we still use it.
* Use `execFile` instead of `exec` to run the compiler, as `exec` spawns a new shell (!!!!!!)
* Don't import `hardhat/console.sol` in the sample project as it's slow to compile.
* Some hardhat load time optimizations — Mostly requiring things when needed instead of at the top of the file
* Cache the solidity files resolution results — We were reading each file multiple times (!)
* We were reading some files just to check for their existence (!)

WIP: Parts of this require tests.

Depends on https://github.com/NomicFoundation/hardhat/pull/3052

### CODE REVIEW

File: packages/hardhat-core/src/internal/util/glob.ts [hunk 1]
```diff
@@ -12,6 +15,9 @@ export async function glob(
   return files.map(path.normalize);
 }
 
+/**
+ * deprecated
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/util/glob.ts:19]:
```suggestion
 * @deprecated
```

File: packages/hardhat-core/src/internal/artifacts.ts [hunk 2]
```diff
@@ -96,21 +90,32 @@ export class Artifacts implements IArtifacts {
...
+        !f.endsWith(".dbg.json")
+    );
 
     return paths.sort();
   }
 
   public async getBuildInfoPaths(): Promise<string[]> {
-    const paths = await glob(this._buildInfosGlob);
+    const paths = await getAllFilesMatching(
+      path.join(this._artifactsPath, BUILD_INFO_DIR_NAME)
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/artifacts.ts:108]: To exactly match the previous implementation, this should filter by `f.endsWith(".json")`

File: packages/hardhat-core/src/internal/solidity/compilation-job.ts [hunk 3]
```diff
@@ -130,7 +129,8 @@ function mergeCompilationJobs(
     }
   }
 
-  return flatten([...jobsMap.values()]);
+  // Array#flat This method defaults to depth limit 1
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/solidity/compilation-job.ts:132]: This answers my previous question. Wow. I had no idea. Should we add a linter rule about not using `.flat` without an argument or something like that? This seems super error-prone.

File: packages/hardhat-core/src/internal/solidity/resolver.ts [hunk 4]
```diff
@@ -116,23 +128,38 @@ export class Resolver {
...
-        return await this.resolveSourceName(normalizeSourceName(imported));
+      let sourceName: string;
+
+      const isRelativeImport = this._isRelativeImport(imported);
+
+      if (!isRelativeImport) {
+        sourceName = normalizeSourceName(imported);
+      } else {
+        sourceName = await this._relativeImportToSourceName(from, imported);
       }
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/solidity/resolver.ts:139]:
```suggestion
      if (isRelativeImport) {
        sourceName = await this._relativeImportToSourceName(from, imported);
      } else {
        sourceName = normalizeSourceName(imported);
      }
```

File: packages/hardhat-core/src/utils/source-names.ts [hunk 5]
```diff
@@ -193,25 +186,19 @@ export function replaceBackslashes(str: string): string {
  * `p` doesn't exist. `p` MUST be in source name format.
  */
 async function getPathTrueCase(fromDir: string, p: string): Promise<string> {
-  const { trueCasePath } = await import("true-case-path");
-
   try {
-    const tcp = await trueCasePath(p, fromDir);
+    const absolute = path.join(fromDir, p);
+    const tcp = await getRealCase(absolute);
```
Comment (fvictorio) [inline packages/hardhat-core/src/utils/source-names.ts:191]: This variable name is not very descriptive now that we are not using `trueCasePath` anymore.

File: packages/hardhat-core/src/internal/util/fs-utils.ts [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
+    throw new FileSystemAccessError(e.message, e);
+  }
+
+  const results = await Promise.all(
+    dir.map(async (file) => {
+      const absolutePathToFile = path.join(absolutePathToDir, file);
+      const stats = await fsPromises.stat(absolutePathToFile);
+      if (stats.isDirectory()) {
+        return getAllFilesMatching(absolutePathToFile, matches);
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/util/fs-utils.ts:106]: Nice finding! 

I'm fine with that error. Anyone creating a symlink loop should burn in hell anyway

File: packages/hardhat-core/src/internal/util/fs-utils.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+      } else {
+        if (matches === undefined || matches(absolutePathToFile)) {
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/util/fs-utils.ts:108]: Disclaimer: this is an after-the-fact justification.

I'd rather not create a new function on each call as this is highly recursive and we know that it's called a TON of times.

File: packages/hardhat-core/src/internal/util/glob.ts [hunk 8]
```diff
@@ -3,6 +3,9 @@ import type { IOptions as GlobOptions } from "glob";
 import * as path from "path";
 import util from "util";
 
+/**
+ * @deprecated
+ */
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/util/glob.ts:8]: The truffle plugin uses it. My fault 😭

I'll add a comment

File: packages/hardhat-core/src/internal/solidity/compilation-job.ts (hunk 3 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/solidity/compilation-job.ts:132]: That'd be useful, but I can't find any rule for that

Comment (alcuadrado) [reply packages/hardhat-core/src/internal/solidity/compilation-job.ts:132]: Slightly OT: https://developer.chrome.com/blog/smooshgate/

File: packages/hardhat-core/sample-projects/javascript/contracts/Lock.sol [hunk 9]
```diff
@@ -21,7 +21,7 @@ contract Lock {
     }
 
     function withdraw() public {
-        // Uncomment this line to print a log in your terminal
+        // Uncomment this line and the import to "hardhat/console.sol" to print a log in your terminal
```
Comment (feuGeneA) [inline packages/hardhat-core/sample-projects/javascript/contracts/Lock.sol:24]:
```suggestion
        // Uncomment this line, and the import of "hardhat/console.sol", to print a log in your terminal
```

File: packages/hardhat-core/src/internal/artifacts.ts (hunk 2 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/artifacts.ts:108]: done

File: packages/hardhat-core/src/utils/source-names.ts (hunk 5 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/utils/source-names.ts:191]: right. renamed it.

File: packages/hardhat-core/src/internal/util/fs-utils.ts [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+/**
+ * Sync version of getRealCase
+ *
+ * @see getRealCase
+ */
+export function getRealCaseSync(absolutePath: string): string {
+  try {
+    return fs.realpathSync.native(absolutePath);
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/util/fs-utils.ts:39]: done

**Comment by alcuadrado** [comment]
Finally ready! This should have been a few prs.

Handling different filesystems at the same time is so painful!

File: packages/hardhat-core/test/internal/core/config/config-loading.ts [hunk 11]
```diff
@@ -334,13 +342,19 @@ Hardhat plugin instead.`
 
         const files = ctx.getFilesLoadedDuringConfig();
 
+        const filesJson = JSON.stringify(files, undefined, 2);
+
         for (const file of builtinTasksFiles) {
-          // The task names may have been loaded before, so we ignore it.
-          if (file.endsWith("task-names.ts")) {
+          // The task names and the utils may have been loaded before, so we ignore them.
+          if (file.endsWith("task-names.ts") || file.includes("utils")) {
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/core/config/config-loading.ts:349]: Maybe?

```suggestion
          if (file.endsWith("task-names.ts") || file.includes(path.join(path.sep, "utils", path.sep))) {
```

**Pull request merged**